lib/Basics/FixedSizeAllocator.h

```diff
@@ -104,63 +97,48 @@ class FixedSizeAllocator {
   FixedSizeAllocator(FixedSizeAllocator const&) = delete;
   FixedSizeAllocator& operator=(FixedSizeAllocator const&) = delete;
 
-  explicit FixedSizeAllocator(size_t itemSize)
-      : _itemSize(itemSize), _freelist(nullptr) {
-    _blocks.reserve(4);
-
-#ifndef TRI_UNALIGNED_ACCESS
-    // align _itemSize to a multiple of sizeof(void*) on system that require it
-    _itemSize +=
-        (_itemSize % sizeof(void*) == 0 ? 0 : sizeof(void*) - _itemSize % sizeof(void*));
-#endif
-  }
-
+  FixedSizeAllocator() = default;
   ~FixedSizeAllocator() = default;
 
-  void* allocate() {
-    if (_freelist != nullptr) {
-      void* element = _freelist;
-      _freelist = *reinterpret_cast<void**>(_freelist);
-      return element;
-    }
-
+  template <typename... Args>
+  T* allocate(Args&&... args) {
     if (_blocks.empty() || _blocks.back()->full()) {
       allocateBlock();
     }
     TRI_ASSERT(!_blocks.empty());
     TRI_ASSERT(!_blocks.back()->full());
-
-    return _blocks.back()->next();
+   
+    try {
+      T* p = _blocks.back()->nextSlot();
+      return new (p) T(std::forward<Args>(args)...);
+    } catch (...) {
+      _blocks.back()->rollbackSlot();
+      throw;
+    }
   }
 
-  void deallocateAll() {
+  void clear() {
     _blocks.clear();
-    _freelist = nullptr;
-  }
-
-  void deallocate(void* value) noexcept {
-    *reinterpret_cast<void**>(value) = _freelist;
-    _freelist = value;
   }
 
-  size_t memoryUsage() const {
-    size_t total = 0;
-    for (auto const& it : _blocks) {
-      total += it->memoryUsage();
+  /// @brief return the total number of used elements, in all blocks
+  size_t numUsed() const noexcept {
+    size_t used = 0;
+    for (auto const& block : _blocks) {
+      used += block->numUsed();
     }
-    return total;
+    return used;
   }
 
  private:
   void allocateBlock() {
-    size_t const size = 128 << (std::min)(size_t(8), _blocks.size());
-    auto block = std::make_unique<MemoryBlock>(_itemSize, size);
-    _blocks.emplace_back(std::move(block));
+    // minimum block size is for 64 items
+    // maximum block size is for 4096 items
+    size_t const numItems = 64 << std::min<size_t>(6, _blocks.size() + 1);
+    _blocks.emplace_back(std::make_unique<MemoryBlock>(numItems));
   }
 
   std::vector<std::unique_ptr<MemoryBlock>> _blocks;
```

Comment:
not that you should change this now, but I wonder whether you could have used std::deque here because I think it sort of does the same thing

Reply:
I would directly link the blocks via next pointers (i.e., create a singly linked list) where head points to the latest block. And I would allocate a larger block memory to contain the `MemoryBlock` together with its items (so only a single allocation is required). I have already made a mental note to create an according PR. ;)

Reply:
These are both good ideas, which will be most helpful for the first few MemoryBlocks.
The current implementation in this PR will do one allocation for the MemoryBlock, one for its data array, and another one for the _blocks vector.
Same for the second block. It will eventually go down to 2 allocations on average, because the vector will double its capacity every time.
But still very useful for massive ASTs!

Reply:
@mpoeter : merging this now. Looking forward to seeing an improving PR from you! Thanks!